src/internal/attr.c

```diff
@@ -0,0 +1,189 @@
+#include <rlang.h>
+#include "internal.h"
+
+static inline sexp* r_language_names(sexp* x);
+static inline sexp* r_names_dispatch(sexp* x);
+
+sexp* rlang_names2(sexp* x) {
+  const enum r_type type = r_typeof(x);
+
+  if (type == r_type_environment) {
+    r_abort("Use `env_names()` for environments.");
+  }
+
+  // Handle pairlists and language objects specially like `getAttrib()`
+  // does. `r_names()` will not find these names because it has a guarantee
+  // to never allocate.
+  if (type == r_type_pairlist || type == r_type_call) {
+    return r_language_names(x);
+  }
+
+  sexp* nms;
+  if (r_is_object(x)) {
+    nms = KEEP(r_names_dispatch(x));
+  } else {
+    nms = KEEP(r_names(x));
+  }
+
+  if (r_is_null(nms)) {
+    r_ssize n = r_length(x);
+    nms = KEEP(r_new_vector(r_type_character, n));
+    r_chr_fill(nms, r_empty_str, n);
+  } else {
+    nms = KEEP(rlang_replace_na(nms, r_shared_empty_chr));
+  }
+
+  FREE(2);
+  return nms;
+}
+
+static inline sexp* r_language_names(sexp* x) {
+  r_ssize n = r_length(x);
+
+  sexp* out = KEEP(r_new_vector(r_type_character, n));
+  sexp** p_out = STRING_PTR(out);
+
+  int i = 0;
+
+  for(; x != r_null; x = r_node_cdr(x), ++i) {
+    sexp* tag = r_node_tag(x);
+
+    if (tag == r_null) {
+      p_out[i] = r_empty_str;
+    } else if (r_is_symbolic(tag)) {
```

Comment:
`r_is_symbolic()` returns true for either symbols and calls. This should be `r_typeof(x) == r_type_symbol`.

But I think this branch and the next one can be removed and you can assume a `TAG` is either `NULL` or a symbol. It's an internal error otherwise, like taking the CDR of a non-node object.

Reply:
The `is_symbol` check came from looking at what base R did, but ill remove
https://github.com/wch/r-source/blob/3d8c9fcaa927eedf0ec77738b1492030aebf75eb/src/main/attrib.c#L124-L145